### PR TITLE
Sync and Release automation

### PR DESCRIPTION
**Pull request by flip1995**
Based on https://github.com/rust-lang/rust-clippy/pull/13693

Adds 2 subcommands to `cargo dev`:

- `cargo dev sync update_nightly`: Which updates the nightly versions in `rust-toolchain` and `clippy_utils/README.md`
- `cargo dev release bump_version`: Bumps the version in all relevant `Cargo.toml` files

Those are pulled out of https://github.com/rust-lang/rust-clippy/pull/12759, which I'll rebase on this.

Next step is to update the documentation, which I'll partly pull out of https://github.com/rust-lang/rust-clippy/pull/12762

r? @blyxyas (as you reviewed the first PR in the chain and were assigned to the second one)

cc https://github.com/rust-lang/rust-clippy/issues/13556

changelog: none

### CODE REVIEW

File: clippy_utils/README.md
Comment (blyxyas) [inline clippy_utils/README.md:1]: This file and `clippy_utils/Cargo.toml` will be changed in #13693

**Comment by blyxyas** [comment]
Seems that there is a CI error, but this has to be the cleanest commit history I've seen in the repo to date. Great work! :cowboy_hat_face:

File: clippy_dev/src/release.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+use std::fmt::Write;
+use std::path::Path;
+
+use crate::utils::{UpdateMode, clippy_version, replace_region_in_file};
+
+const CARGO_TOML_FILES: [&str; 5] = [
```
Comment (BD103) [inline clippy_dev/src/release.rs:6]:
```suggestion
const CARGO_TOML_FILES: [&str; 4] = [
```

This should fix the CI error :)

File: clippy_utils/README.md
Comment (flip1995) [reply clippy_utils/README.md:1]: Yes, this branch is based on the branch of #13693

**Comment by blyxyas** [comment]
Something I'm wondering, what happens with versions like `1.72.1`, would it follow a `0.Major.Minor-Patch` convention?

File: clippy_dev/src/main.rs [hunk 2]
```diff
@@ -291,3 +301,29 @@ enum RemoveSubcommand {
...
+    #[command(subcommand)]
+    subcommand: ReleaseSubcommand,
+}
+
+#[derive(Subcommand)]
+enum ReleaseSubcommand {
+    #[command(name = "bump_version")]
+    /// Bump the version in the Cargo.toml files
+    BumpVersion,
+}
```
Comment (blyxyas) [inline clippy_dev/src/main.rs:329]: Is there a reason for these two to be subcommands in their own structs? Is there a plan to add new commands to them?

**Review by blyxyas — COMMENTED** [review]
Just a question, and this is looking pretty good!

**Comment by flip1995** [comment]
> Something I'm wondering, what happens with versions like `1.72.1`, would it follow a `0.Major.Minor-Patch` convention?

Nah, we won't bump the version of our crates with patch versions. We're also no longer doing patch releases in Clippy. So this is not a concern.

**Review by blyxyas — COMMENTED** [review]
It currently has an error in the second commit (which causes a panic)

File: clippy_dev/src/sync.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+use std::path::Path;
+
+use chrono::offset::Utc;
+
+use crate::utils::{UpdateMode, replace_region_in_file};
+
+pub fn update_nightly() {
+    // Update rust-toolchain nightly version
+    let date = Utc::now().format("%Y-%m-%d").to_string();
+    replace_region_in_file(
```
[the remark this answers is not in the document]
Comment (flip1995) [reply clippy_dev/src/sync.rs:11]: Oh good catch. That must've been lost during a rebase. Added those in the sync commit. Nothing else changed in the force push.

**Review by blyxyas — APPROVED** [review]
LGTM, thanks! :heart: :rainbow:

**Pull request merged**